File: public/dsa/linked_lists/reverse_a_linked_list/iterative/reverse_a_linked_list.cpp

#include <bits/stdc++.h>
using namespace std;

// defs-start
class ListNode {
   public:
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};

// defs-end

ListNode* reverseLinkedList(ListNode* head) {
    ListNode* cur = head;
    ListNode* prev = nullptr;

    while (cur) {
        ListNode* next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }

    return prev;
}

// tests-start
int main() {
    ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));

    ListNode* res = reverseLinkedList(head);

    ListNode* temp = res;
    while (temp) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;

    return 0;
}
// tests-end

/*output
4 3 2 1
*/

File: public/dsa/linked_lists/reverse_a_linked_list/iterative/reverse_a_linked_list.py

# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def reverseLinkedList(self, head: ListNode) -> ListNode:
        cur = head
        prev = None
        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        return prev


# tests-start
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
sol = Solution()
res = sol.reverseLinkedList(head)

temp = res
while temp:
    print(temp.data, end=' ')
    temp = temp.next

# tests-end


'''output
4 3 2 1
'''
